apps/dashboard/src/components/primitives/field-editor/field-editor.tsx

```diff
@@ -0,0 +1,232 @@
+
+type SelectedVariable = {
+  value: string;
+  from: number;
+  to: number;
+} | null;
+
+type CompletionRange = {
+  from: number;
+  to: number;
+} | null;
+
+type FieldEditorProps = {
+  value: string;
+  onChange: (value: string) => void;
+  variables: LiquidVariable[];
+  placeholder?: string;
+  autoFocus?: boolean;
+  size?: 'default' | 'lg';
+  fontFamily?: 'inherit';
+  id?: string;
+};
+
+export function FieldEditor({
+  value,
+  onChange,
+  variables,
+  placeholder,
+  autoFocus,
+  size = 'default',
+  fontFamily = 'inherit',
+  id,
+}: FieldEditorProps) {
+  const viewRef = useRef<EditorView | null>(null);
+  const lastCompletionRef = useRef<CompletionRange>(null);
+
+  const { selectedVariable, setSelectedVariable, handleVariableSelect, isUpdatingRef } = useVariableSelection();
+
+  const handleVariableUpdate = useVariableUpdate(
+    selectedVariable,
+    viewRef,
+    isUpdatingRef,
+    onChange,
+    setSelectedVariable
+  );
+
+  const completionSource = useCompletionSource(variables, lastCompletionRef);
+
+  const extensions = useMemo(
+    () => [
+      autocompletion({
+        override: [completionSource],
+        closeOnBlur: false,
+        defaultKeymap: true,
+        activateOnTyping: true,
+      }),
+      EditorView.lineWrapping,
+      variablePillTheme,
+      createVariablePlugin({ viewRef, lastCompletionRef, onSelect: handleVariableSelect }),
+    ],
+    [variables, completionSource, handleVariableSelect]
+  );
+
+  return (
+    <div className="relative">
+      <Editor
+        size={size}
+        className="flex-1"
+        autoFocus={autoFocus}
+        fontFamily={fontFamily}
+        placeholder={placeholder}
+        id={id}
+        extensions={extensions}
+        value={value}
+        onChange={onChange}
+      />
+      <Popover
+        open={!!selectedVariable}
+        onOpenChange={(open) => {
+          if (!open) {
+            setTimeout(() => setSelectedVariable(null), 0);
+          }
+        }}
+      >
+        <PopoverTrigger asChild>
+          <div />
+        </PopoverTrigger>
+        {selectedVariable && (
+          <VariablePopover
+            variable={selectedVariable.value}
+            onClose={() => setSelectedVariable(null)}
+            onUpdate={handleVariableUpdate}
+          />
+        )}
+      </Popover>
+    </div>
+  );
+}
+
+/**
+ * Manages the state and selection of Liquid variables in the editor.
+ *
+ * This hook handles the complex state management needed when a user selects a variable in the editor:
+ * 1. Tracks which variable is currently selected
+ * 2. Prevents recursive updates when variables are being modified
+ */
+function useVariableSelection() {
+  const [selectedVariable, setSelectedVariable] = useState<SelectedVariable>(null);
+  const isUpdatingRef = useRef(false);
+
+  const handleVariableSelect = useCallback((value: string, from: number, to: number) => {
+    if (isUpdatingRef.current) return;
+
+    requestAnimationFrame(() => {
```

Comment:
`requestAnimationFrame` needs a clean-up function on unmount, similar to setTimeout. See [here](https://css-tricks.com/using-requestanimationframe-with-react-hooks/)